### ISSUE TITLE
regression: Sudden OOM in the pg-cdc-snapshot scenario of the bounded-memory test

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#22274]
### What version of Materialize are you using?

87f62832e87a30c4889e8568a5475a98efd1a336

### What is the issue?

A workload that has not given any trouble recently has started failing reliably in Nightly.

To reproduce:

1. Use a `m6a.8xlarge` EC2 instance with the `ami-02c624694950f12f2` AMI
2. Run
```
bin/ci-builder run stable bin/mzcompose --find bounded-memory run default pg-cdc-snapshot
```

and you will get:

```
> SELECT * FROM v1; /* expect 2097152 */
rows didn't match; sleeping to see if dataflow catches up 50ms 75ms 113ms 169ms 253ms 380ms 570ms 854ms 1s 2s 3s ...
```

at which point `journalctl` will report that the OOM killer has activated to kill a `clusterd`:

```
Oct 10 11:17:35 ip-172-31-92-39.ec2.internal kernel: Memory cgroup out of memory: Killed process 77949 (clusterd) total-vm:5133168kB, anon-rss:1610272kB, file-rss:63352kB, shmem-rss:0kB, UID:999 pgtables:6176kB oom_score_adj:0
```

`docker ps -a` will report that the environmentd has exited:

```
dd4f6f28e681   materialize/materialized:mzbuild-X5SYOTVCX32VIIXEPB3ESYQ3LZI4MITJ   "tini -- entrypoint.…"   About a minute ago   Exited (137) 14 seconds ago                                                                                                                                           bounded-memory-materialized-1
```

Bisection points to this commit:

```
commit 87f62832e87a30c4889e8568a5475a98efd1a336
Author: Jan Teske <jteske@posteo.net>
Date:   Fri Sep 8 19:29:20 2023 +0200

    compute: make error logging work with reused indexes
```

which changes the definition of a builtin view from a simple projection to a `WITH MUTUALLY RECURSIVE`. However, this commit arrived in a group of related commits so any of them could be to blame.

**Comment by philip-stoev** [comment]
More detailed bisection now in progress ...

**Comment by philip-stoev** [comment]
I was unable to bisect it any better because git bisect gets confused.

What I tried is giving more memory to the test. I had to give it 400Mb more to make it pass:

```
diff --git a/test/bounded-memory/mzcompose.py b/test/bounded-memory/mzcompose.py
index 537a496fdd..82dd8a36f7 100644
--- a/test/bounded-memory/mzcompose.py
+++ b/test/bounded-memory/mzcompose.py
@@ -25,8 +25,8 @@ PAD_LEN = 1024
 STRING_PAD = "x" * PAD_LEN
 REPEAT = 16 * 1024
 ITERATIONS = 128
-MATERIALIZED_MEMORY = "5Gb"
-CLUSTERD_MEMORY = "3.5Gb"
+MATERIALIZED_MEMORY = "5.4Gb"
+CLUSTERD_MEMORY = "3.9Gb"
```

**Comment by teskje** [comment]
Unfortunately, I wasn't able to make any progress here. I was able to repro the OOM on main with the instructions given in the description, but I was also able to repro it:

* On the commit before https://github.com/MaterializeInc/materialize/pull/21640 was merged (64dc2f1c2b119894e6a5cd1f8ac963610183e7f9)
* On a branch based on main that had https://github.com/MaterializeInc/materialize/pull/21640 reverted

Finally, I also attempted to reproduce the pg-cdc-snapshot test case locally and take memory profiles of its execution, but the profiles looked the same with and without https://github.com/MaterializeInc/materialize/pull/21640, and maximum memory usage was in the megabytes instead of gigabytes, which makes me think that my reproduction wasn't at all accurate.

So there is nothing to suggest that https://github.com/MaterializeInc/materialize/pull/21640 is to blame here, but also no evidence to the contrary. If no dataflow errors are produced, https://github.com/MaterializeInc/materialize/pull/21640 _shouldn't_ increase memory usage but of course there might be a bug somewhere.

**Comment by philip-stoev** [comment]
In Nightly, this only happened for `3479d1bd501a2be01c5e838cbe577a02159c9bd7` and never again.
